### ISSUE TITLE
need equivalent of --ntasks-per-node

### ISSUE THREAD

**Issue by grondo** [flux-framework/flux-core#4533]
I don't believe there is an easy way to request a number of nodes and set the number of tasks to run on those nodes outside of the shell-provided backdoor `-o per-resource.type=node -o per-resource.count=N`. If it is possible to make this request with jobspec v1, we should consider how to add this option to `flux mini run`. If not, then maybe hiding the shell option w/a behind such a `flux mini run` option would be better than telling users to set those options directly.

**Comment by trws** [comment]
I can't remember entirely how our `per-resource` support is currently implemented, but jobspec V1 includes support for `per_slot` in `count` rather than `total`, so we could construct a `nodes[count]>slot` resource spec and then use `count:per_slot:tasks_per_node` to get there.  We could probably also do the math locally as long as we require the number of nodes to be explicit, and only allow exclusive nodes with that option, then `count:total:count * tasks_per_node` should do it.

**Comment by grondo** [comment]
> I can't remember entirely how our per-resource support is currently implemented,

Sadly, this option causes the job shell to completely ignore the `tasks` section of jobspec and do exactly what the option states, run N tasks per resource type noted in the option (only nodes, cores supported at this time)

> so we could construct a nodes[count]>slot resource spec and then use count:per_slot:tasks_per_node to get there.

I'm probably wrong, but [RFC 25](https://flux-framework.readthedocs.io/projects/flux-rfc/en/latest/spec_25.html#v1-specific-resource-graph-restrictions) lists the only allowed specs as:

 * slot>core
 * node>slot>core
 * slot>(core,gpu)
 * node>slot>(core,gpu)

i.e. `slot` always has to contain a `core` and/or `gpu`. If we could relax this constraint, then I can see how `node>slot` could be used with `count:per_slot` (note that the job shell currently does not support `count:per_slot` afaik, so that would need to be addressed as well). I'm not sure why we limited the spec as it is now though...

**Comment by trws** [comment]
Ah, then the easiest is probably to make `--ntasks-per-node` generate the `per-resource` form with `resource.type=node` and just prevent them from supplying other task options.  Would that work or cause other issues do you think?

**Comment by grondo** [comment]
That would work, and is a good solution since we can change how it is implemented in jobspec without breaking the "interface"

**Comment by grondo** [comment]
> and just prevent them from supplying other task options. 

This is actually a little tricky. We could easily prevent use of `-n, --ntasks`, but then there is no way to request a number of cores, since the number of cores is set by `ntasks * cores_per_task` in the current `Jobspec.from_command()` function. Maybe that is fine, since users making use of `--ntasks-per-node=N` would most likely also be using `-N` and not `-n`. However, a nice use case here might be to launch a new flux instance (allocate 24 cores with one flux start per node)

```
flux mini run -n 24 --ntasks-per-node=1 flux start ...
```

Maybe we can just say that `--ntasks-per-node` "Forces N tasks per node regardless of other task specification options"? And forgo the stringent option checking necessary otherwise?

Or, maybe there is another better way to provide this functionality.

**Comment by trws** [comment]
We could, but the first question that comes to mind is why can’t we add the `—cores` option so we don’t have to conflate tasks with cores in the example?





On September 2, 2022 at 3:26:30 PM PDT, Mark Grondona ***@***.***> wrote:

and just prevent them from supplying other task options.

This is actually a little tricky. We could easily prevent use of -n, --ntasks, but then there is no way to request a number of cores, since the number of cores is set by ntasks * cores_per_task in the current Jobspec.from_command() function. Maybe that is fine, since users making use of --ntasks-per-node=N would most likely also be using -N and not -n. However, a nice use case here might be to launch a new flux instance (allocate 24 cores with one flux start per node)

flux mini run -n 24 --ntasks-per-node=1 flux start ...


Maybe we can just say that --ntasks-per-node "Forces N tasks per node regardless of other task specification options"? And forgo the stringent option checking necessary otherwise?

Or, maybe there is another better way to provide this functionality.

**Comment by grondo** [comment]
Good question, but I don't remember why the command and `Jobspec` class were designed this way.

https://github.com/flux-framework/flux-core/blob/a34a2292c4a08e2f211d41b61aa2debe645701fb/src/bindings/python/flux/job/Jobspec.py#L652-L672

**Comment by trws** [comment]
That’s just the constructor function that takes a command though right? What it generates can include arbitrary cores, and in fact there is at least one case where it produces a per_slot task count. Should be possible to extend that in either way we want, though it will take some care to keep it backwards compatible.


---
Sent from Workspace ONE Boxer<https://whatisworkspaceone.com/boxer>

On September 2, 2022 at 3:59:30 PM PDT, Mark Grondona ***@***.***> wrote:

Good question, but I don't remember why the command and Jobspec class were designed this way.

https://github.com/flux-framework/flux-core/blob/a34a2292c4a08e2f211d41b61aa2debe645701fb/src/bindings/python/flux/job/Jobspec.py#L652-L672<https://urldefense.us/v3/__https://github.com/flux-framework/flux-core/blob/a34a2292c4a08e2f211d41b61aa2debe645701fb/src/bindings/python/flux/job/Jobspec.py*L652-L672__;Iw!!G2kpM7uM-TzIFchu!j8E9uYxLB52P5DdndeezSFcKYQ2qycXxSPN3QFJNCkVZ5vfqsbHfKHGqz3XWkBXwBA$>

**Comment by grondo** [comment]
@trws: after I attempting to incorporate `--tasks-per-node` and `--cores` into the existing `JobspecV1.from_command()` constructor, I quickly found that doing so while preserving guaranteed backwards compatibility, readable and understandable code, and an implementation that didn't drive someone crazy was a losing battle.

Instead, I propose adding a new constructor which is used if any of a set of new options are passed with `flux mini run` or `submit`:

```
Per resource options:
  The following options allow per-resource specification of tasks, and
  should not be used with per-task options above

      --cores=N             Request a total number of cores
      --tasks-per-node=N    Force number of tasks per node
      --tasks-per-core=N    Force number of tasks per core
      --gpus-per-node=N     Request a number of GPUs per node. Only with
                            --nnodes, --tasks-per-node
```

This lets us separate the code for these two use cases, making the interfaces more understandable and testable. For now, the "per resource" interface is somewhat limited by design, since we're mainly trying to satisfy a limited use case at this time.

There are more details in PR #4544. Can you take a look today and let us know what you think? We're going to tag today and would like to know sooner rather than later if we should incorporate this change into the release, or delay for further discussion. Thanks!

**Comment by trws** [comment]
That sounds like a very good idea. I’ll take a look at the PR momentarily.

**Comment by grondo** [comment]
This should have been closed by #4544
